### PR TITLE
Add SGDOptimizer in the on-device training offline tooling (onnxblock)

### PR DESCRIPTION
**Pull request by AdamLouly**
### Description
Adding SGDOptimizer to on device training onnxblock

### CODE REVIEW

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 1]
```diff
@@ -211,3 +211,82 @@ def build(self, parameters):
...
+        self._sgd = SGDOptimizer(learning_rate)
+
+    def build(self, parameters):
+        """Returns an SGD optimizer model based on the input parameters."""
+
+        # get the model to manipulate and update its namespace
+        onnx_model = self.base
+
+        # TODO: Avoid hard coded input/output strings
+        learning_rate_name = "lr"
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:268]: where are these hard coded names used? In the c++ optimizer.cc?

Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:268]: here
https://github.com/microsoft/onnxruntime/blob/c424e42594d92daba54f264c1c7409e53529d933/orttraining/orttraining/core/graph/training_op_defs.cc#L1299

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 2]
```diff
@@ -211,3 +211,82 @@ def build(self, parameters):
...
+    Creates a block that updates the model parameters based on the calculated
+    gradient following the SGD algorithm.
+
+    Args:
+        learning_rate: float indicating the learning rate.
+    Returns:
+        Returns a string of the output names from this optimizer node.
+    """
+
+    def __init__(self, learning_rate: float):
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:257]: Enable gradient normalization?

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py [hunk 3]
```diff
@@ -554,6 +554,83 @@ def mse_loss(prediction, target):
         _ = ort_session.run(ort_output_names, ort_inputs)
 
 
+@pytest.mark.parametrize(
+    "block",
+    [SimpleTrainingBlockWithMSELoss, SimpleTrainingBlockWithCrossEntropyLoss, SimpleTrainingBlockWithBCEWithLogitsLoss],
+)
+def test_sgd_optimizer_composition(block):
+    # Given
+    device = "cuda"
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_onnxblock.py:563]: Since this is only composition test we can just use cpu device.
BTW curious are these tests only enabled in GPU pipelines?

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py [hunk 4]
```diff
@@ -554,6 +554,83 @@ def mse_loss(prediction, target):
...
+        _ = optimizer(simple_block.parameters())
+        optimizer_model = accessor.model
+        assert optimizer_model
+
+
+# TODO: Add a test for correctness when creation of ortvalues of
+# tensor seq is possible on cuda
+def test_sgd_optimizer_execution():
+    # Given
+    device = "cuda"
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_onnxblock.py:587]: why are we only testing on cuda? This is for on-device training.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 1 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:268]: These are graph input names. They need not match the schema names.

They should match the names expected by the `Optimizer`:

https://github.com/microsoft/onnxruntime/blob/31cbd63af7a8a8582865882b1870719724b6e9b4/orttraining/orttraining/training_api/optimizer.cc#L22

Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:268]: @pengwa do you happen to know why `step` is not an input to the SGDOptimizerV2?
From optimizer.cc, it seems to be a required input.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 5]
```diff
@@ -211,3 +211,82 @@ def build(self, parameters):
         )
 
         return updated_flag_name
+
+
+class SGDOptimizer(blocks.Block):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:216]: There seems to be quite an overlap between SGDOptimizer and AdamWOptimizer.
Can we factor out the common logic into a parent class `_Optimizer` or `_OptimizerBase` which gets called from both of these optimizer's build method?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:216]: You could do something similar for the `AdamW` and `SGD` forward blocks as well.

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py [hunk 6]
```diff
@@ -554,6 +554,83 @@ def mse_loss(prediction, target):
...
+    optimizer = onnxblock.optim.SGD(learning_rate)
+    with onnxblock.empty_base() as accessor:
+        _ = optimizer(simple_block.parameters())
+        optimizer_model = accessor.model
+        assert optimizer_model
+
+
+# TODO: Add a test for correctness when creation of ortvalues of
+# tensor seq is possible on cuda
+def test_sgd_optimizer_execution():
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_onnxblock.py:585]: I wrote a lot of these tests before we had the python bindings.
Can we leverage the python bindings now that we have them to tell whether the graph can be executed correctly using the training api? Maybe you can move this test to `orttraining_test_python_bindings`.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 5 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:216]: Creating a shared base class for the Optimizer Classes is a viable option. 
However, for the AdamW and SGD classes, their high level of customization, which is closely tied to their specific parameters and node attributes, might not be that useful and will make that class complicated.

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py (hunk 4 above)
Comment (AdamLouly) [reply orttraining/orttraining/test/python/orttraining_test_onnxblock.py:587]: will change to cpu

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py (hunk 3 above)
Comment (AdamLouly) [reply orttraining/orttraining/test/python/orttraining_test_onnxblock.py:563]: Seems like they only run on Test Training APIs which is on GPU.
however, we will change it to the on device training tests instead of onnxblock.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:268]: I believe SGD don't need step count (for tuning learning rate). Then I think I should not put step in  CommonOptimizerInputs earlier... Only add it when needed (in adamw)

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 7]
```diff
@@ -125,77 +60,207 @@ def build(self, gradients_name: str):
         return cgn_node_output_name
 
 
-class AdamW(onnxblock_module.ForwardBlock):
-    """Builds AdamW optimizer onnxblock for the given training parameters.
+class _OptimizerBase(blocks.Block):
+    def __init__(self):
+        super().__init__()
+
+    def build_optimizer_node(self, input_names, output_name, node_name, node_domain, node_attributes):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:67]: add type hints?

File: onnxruntime/test/testdata/training_api/adamw.onnx
Comment (baijumeswani) [inline onnxruntime/test/testdata/training_api/adamw.onnx:1]: just for my knowledge, why was this file regenerated?

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 7 above)
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:67]: Why not call this function `build` and invoke it by calling super().build(...)?

File: orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc [hunk 8]
```diff
@@ -35,7 +35,7 @@ Status AdamWOptimizer::ComputeInternal(OpKernelContext* ctx) const {
   AdamWOptimizerBase::Prepare p;
   ORT_RETURN_IF_ERROR(PrepareForCompute(ctx, p));
 
-  int64_t* updated_flag_ptr = p.updated_flag->template MutableData<int64_t>();
+  bool* updated_flag_ptr = p.updated_flag->template MutableData<bool>();
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc:38]: Why is this change made?

Comment (AdamLouly) [reply orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc:38]: This change was needed because the code was originally checking if the output updated_flag was an integer, but in the SGD optimizer, it was defined as a boolean. Instead of adding a new check for booleans, I made the definition of adamW match the boolean type, and then changed the check to only look for booleans.

File: onnxruntime/test/testdata/training_api/adamw.onnx
Comment (AdamLouly) [reply onnxruntime/test/testdata/training_api/adamw.onnx:1]: since we chanted the output type to bool, regenerating this was needed to create a new graph with output type bool

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 7 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:67]: just to make the code more self explanatory, can change it if its crucial.

File: orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc (hunk 8 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc:38]: This is a legacy todo, SGD and AdamW define updated_flag in different data types.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 9]
```diff
@@ -125,77 +60,231 @@ def build(self, gradients_name: str):
...
-    """Builds AdamW optimizer onnxblock for the given training parameters.
+class _OptimizerBase(blocks.Block):
+    def __init__(self):
+        super().__init__()
 
-    Creates a block that updates the model parameters based on the calculated
-    gradient following the AdamW algorithm.
+    def build_optimizer_node(
+        self,
+        input_names: list,
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:69]:
```suggestion
        input_names: list[str],
```

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 10]
```diff
@@ -205,9 +294,34 @@ def build(self, parameters):
...
+    def __init__(self, clip_grad=None):
+        super().__init__(clip_grad)
+        self._sgd = SGDOptimizer()
+
+    def _optimizer_specific_logic(
+        self,
+        learning_rate_name: str,
+        params_name: str,
+        gradients_name: str,
+        trainable_parameters: Tuple,
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:317]:
```suggestion
        trainable_parameters: Tuple[???],
```

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 11]
```diff
@@ -125,77 +60,231 @@ def build(self, gradients_name: str):
...
 
-class AdamW(onnxblock_module.ForwardBlock):
-    """Builds AdamW optimizer onnxblock for the given training parameters.
+class _OptimizerBase(blocks.Block):
+    def __init__(self):
+        super().__init__()
 
-    Creates a block that updates the model parameters based on the calculated
-    gradient following the AdamW algorithm.
+    def build_optimizer_node(
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:67]:
```suggestion
    def _build_optimizer_node(
```

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 12]
```diff
@@ -125,77 +60,231 @@ def build(self, gradients_name: str):
...
+            node_name (str): Name of the optimizer node.
+            node_domain (str): Domain of the optimizer node.
+            node_attributes (dict): Additional attributes for the optimizer node.
+
+        Returns:
+            str: The output tensor name of the optimizer node.
+        """
+        onnx_model = self.base
+        optimizer_input_names = input_names
+        optimizer_output_names = [output_name]
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:90]: You can avoid creating these intermediate variables.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 13]
```diff
@@ -125,77 +60,231 @@ def build(self, gradients_name: str):
...
+            str: The output tensor name of the SGD optimizer node.
+        """
+
+        input_names = [learning_rate_name, gradients_name, params_name]
+
+        return self.build_optimizer_node(
+            input_names,
+            _graph_utils.generate_graph_name("update_completed"),
+            "SGDOptimizerV2",
+            "com.microsoft",
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:135]: domain name need not be part of the input arguments to the function since it is common and probably all future optimizers will share the same domain.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 14]
```diff
@@ -125,77 +60,231 @@ def build(self, gradients_name: str):
...
+        second_order_moments_name = "second_order_moments"
+
+        # Prepare the tensor sequence inputs for moments
+        onnx_model.graph.input.append(
+            onnx.helper.make_tensor_sequence_value_info(
+                second_order_moments_name, trainable_parameters[0].data_type, None
+            )
+        )
+
+        # ... Prepare tensor sequence inputs for AdamW
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:287]:
```suggestion
```

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 15]
```diff
@@ -205,9 +294,34 @@ def build(self, parameters):
             second_order_moments_name,
         )
 
-        # Create the graph outputs
+        # ... Create graph outputs for AdamW
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:297]:
```suggestion
        # Create graph outputs for AdamW
```

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 16]
```diff
@@ -205,9 +294,34 @@ def build(self, parameters):
...
+        self,
+        learning_rate_name: str,
+        params_name: str,
+        gradients_name: str,
+        trainable_parameters: Tuple[List[onnx.TensorProto], List[onnx.TensorProto]],
+    ) -> str:
+        onnx_model = self.base
+        updated_flag_name = self._sgd(learning_rate_name, params_name, gradients_name)
+
+        # ... Create graph outputs for SGD
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:322]:
```suggestion
        # Create graph outputs for SGD
```

File: orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc:38]: @AdamLouly you probably need to set the `updated_flag_ptr` accordingly (to true or false instead of 1 and 0)

File: orttraining/orttraining/training_ops/cpu/optimizer/adamw/adamw.cc [hunk 17]
```diff
@@ -137,7 +137,7 @@ Status AdamWOptimizer<T>::Compute(OpKernelContext* ctx) const {
   AdamWOptimizerBase::Prepare p;
   ORT_RETURN_IF_ERROR(PrepareForCompute(ctx, p));
 
-  int64_t* updated_flag_ptr = p.updated_flag->template MutableData<int64_t>();
+  bool* updated_flag_ptr = p.updated_flag->template MutableData<bool>();
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cpu/optimizer/adamw/adamw.cc:140]: You need to set this bool flag to `true`/`false`

**Review by baijumeswani — APPROVED** [review]

**Comment by AdamLouly** [comment]
Thanks for the review!
